### PR TITLE
Bounding boxes example

### LINKED ISSUE #247: Add bounding boxes example
**Is your feature request related to a problem? Please describe.**
Since movement is expanding to include bounding boxes in its analyses, we could add an example for a bounding boxes case.

**Describe the solution you'd like**
An example similar to the [Load and explore pose tracks](https://movement.neuroinformatics.dev/examples/load_and_explore_poses.html#sphx-glr-examples-load-and-explore-poses-py) example, for bounding boxes.

**Describe alternatives you've considered**
We could have it in the same notebook as the pose tracks but that feels a bit unnecessarily cramped.

**Additional context**
\

### PR DESCRIPTION
**Pull request by sfmig**
Rebase after #313 is merged

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
To demo the bounding boxes' dataset.

**What does this PR do?**
- It shows how to load a sample bounding boxes dataset and its structure
- It shows how to plot bounding boxes using the position and shape arrays
- It shows different simple options to fill in missing values in the position and shape arrays
- It also fixes a typo in the input/output documentation re exporting bounding boxes datasets as csv files.

## References

Closes #247. 

Rebase after #313 is merged

## How has this PR been tested?

Docs build locally and test pass locally and in CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

No.

## Checklist:

- [x] The code has been tested locally
- [ n/a ] Tests have been added to cover all new functionality
- [ n/a ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by lochhh** [comment]
> ## [Quality Gate Passed](https://sonarcloud.io/dashboard?id=neuroinformatics-unit_movement&pullRequest=312) **Quality Gate passed**
> Issues  [3 New issues](https://sonarcloud.io/project/issues?id=neuroinformatics-unit_movement&pullRequest=312&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) 

Need resolving

**Comment by sfmig** [comment]
hey, thanks for having a look!

From the issues flagged by sonarcloud:
- two are due to duplicate lines, because I am printing explanatory lines a few times in the example. Namely:
    - `print("Position data array (first 14 frames):")` and
    - `print("Shape data array (first 14 frames):")`.
- one is due to commented out code for `# sphinx_gallery_thumbnail_number = 2`, but this is the syntax to change the default thumbnail shown in the example. We are using the same approach in [another example](https://github.com/neuroinformatics-unit/movement/blob/85904d9c2a4fd65f1e78d046f9f869ba83ee88af/examples/compute_polar_coordinates.py#L304) too.

I think for clarity it makes sense to keep the duplicate lines as is and not define a separate variable, but lmk if you disagree.

I cannot retrieve the reports from previous tutorials anymore (examples [here](https://github.com/neuroinformatics-unit/movement/pull/180#issuecomment-2129496480) and [here](https://github.com/neuroinformatics-unit/movement/pull/135#issuecomment-2007119996)) but it seems in those cases we skipped three issues too.

File: examples/load_and_reindex_bboxes.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
+"""Load and reindex bounding boxes tracks
+==========================================
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:2]: I'm unsure about including the term "reindex" in the title of this example. If one reads it out of context, e.g. a user looking at the gallery of examples, it doens't convey much information on what is being done, because "indexing" is a very loaded term that carries many meanings. You are probably referring to the `xarray` method of the same name, which has a very specific meaning, both most people won't be familiar with that. 

We should try to convey the meaning more concretely. Perhaps "upsample" could be used here? That's essentially what we are doing right? We are generating more datapoints in the time dimension. So perhaps: "Load and upsample bounding boxes tracks". 
The tagline (the sentence below the title) can give a bit more info: "Load bounding boxes tracks and upsample them to match the video frame rate". To me that makes it much clearer what this is about, but my view is definitely subjective.

File: examples/load_and_reindex_bboxes.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+# ``position``, ``shape``, and ``confidence``.
+print(ds)
+
+# %%
+# We can see the coordinates in the time dimension are expressed in frames,
+# and that we only have data for 1 in 5 frames of the video, plus
+# the last frame (167).
+#
+# In the following sections of the notebook we will explore options to reindex
+# the dataset and fill in values for the frames with missing data.
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:58]:
```suggestion
# In the following sections of the notebook we will explore options to upsample
# the dataset by filling in values for video frames with no data.
```

Or some variant to the same effect.

File: examples/load_and_reindex_bboxes.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+print(ds_nan.position.data[:14, 0, :])
+
+print("----")
+print("Shape data array (first 14 frames):")
+print(ds_nan.shape.data[:14, 0, :])
+
+# %%
+# Linearly interpolate NaN values
+# ----------------------------------------------------------
+# We can instead fill in the missing values in the dataset applying linear
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:196]:
```suggestion
# We can instead fill in the missing values in the dataset by applying linear
```

File: examples/load_and_reindex_bboxes.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    ds_interp[data_array_str] = interpolate_over_time(
+        data=ds_interp[data_array_str],
+        method="linear",
+        max_gap=None,
+        print_report=False,
+    )
+
+# check the first 14 frames of the data
+print("Position data array (first 14 frames):")
+print(ds_interp.position.data[:14, 0, :])
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:216]:
```suggestion
print(ds_interp.position.isel(time=slice(0, 14), individuals=0).data) 
````

File: examples/load_and_reindex_bboxes.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+print("----")
+print("Shape data array (first 14 frames):")
+print(ds_interp.shape.data[:14, 0, :])
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:220]:
```suggestion
print(ds_interp.shape.isel(time=slice(0, 14), individuals=0).data) 
```

File: examples/load_and_reindex_bboxes.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+
+# %%
+# Compare interpolation methods
+# ------------------------------
+# We can now qualitatively compare the three different methods of filling
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:226]: I was confused by this wording. There are only two methods of filling here, ffill, or linear. 
But I then understood you count filling with NaNs as the third method (essentially mark the empty frames as missing).

Maybe you can clarify the wording a bit to convey what the three variants are.

File: examples/load_and_reindex_bboxes.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    # add subplot axes
+    ax = plt.subplot(3, 2, frame_n + 1)
+
+    # plot frame
+    # note: the video is indexed at every frame, so
+    # we use the frame number as index
+    ax.imshow(video[frame_n])
+
+    # plot bounding box for each dataset
+    for ds_i, ds_one in enumerate([ds_nan, ds_ff, ds_interp]):
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:250]: I was confues by the use of `ds_one`. I would expect `ds_` as a quasi-convention in this case.
`ds_one` made me think of "first dataset".

File: examples/load_and_reindex_bboxes.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+            s=5,
+            color=list_colors[ds_i],
+        )
+
+    # add legend to first frame
+    if frame_n == 0:
+        ax.legend()
+    ax.set_title(f"Frame {frame_n}")
+    ax.set_xlabel("x (pixels)")
+    ax.set_ylabel("y (pixels)")
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:290]: If you adopt any of my suggestions I gave for the earlier plotting function, make sure this one follows the same conventions namings etc, for consistency.

Comment (niksirbi) [reply examples/load_and_reindex_bboxes.py:290]: An aside on this plot, I find the yellow dot (centroid of ffil bbox) quite hard to see against the backdrop of the frame. Perhaps worth experimenting with a different color?

File: examples/load_and_reindex_bboxes.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+# %%
+# Export as .csv file
+# -------------------
+# Let's assume the dataset with the forward filled values is the best suited
+# for our task - we can now export the computed values to a .csv file
+#
+# Note that we currently do not provide explicit methods to export a
+# ``movement`` bounding boxes dataset in a specific format. However, we can
+# easily save the bounding boxes’ trajectories to a .csv file using the
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:302]: Replace the unicode character for the apostrophe. Currently you are using a "right quotation mark" instead of an apostrophe. Or get rid of the appostrophe altogether, as I suggest above.
```suggestion
# easily save the bounding boxes' trajectories to a .csv file using the
```

No, I'm not eagle-eyes,I just have a VSCode extension that flags this ;)

File: examples/load_and_reindex_bboxes.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+# define name for output csv file
+filepath = "tracking_output.csv"
+
+# open the csv file in write mode
+with open(filepath, mode="w", newline="") as file:
+    writer = csv.writer(file)
+
+    # write the header
+    writer.writerow(
+        ["frame_idx", "bbox_ID", "x", "y", "width", "height", "confidence"]
```
Comment (niksirbi) [inline examples/load_and_reindex_bboxes.py:314]: Do we know the fps for this data? If so, we could even add a 'time' column?
Or perhaps this is scope creep and just ignore me.

**Review by niksirbi — APPROVED** [review]
Thank you @sfmig, this is a great example! I left a bunch of comments, many of them nitpicky and/or subjective.

My two substantial points are:
- I think we should change the name and/or tagline of this example. I don't think that "reindexing" means much to most users
- we should rely on selecting the data with `sel` and `isel` as much as possible, and avoid numpy-style indexing (which is contingent on a certain dimension order).

**Comment by niksirbi** [comment]
> hey, thanks for having a look!
> 
> From the issues flagged by sonarcloud:
> 
> * two are due to duplicate lines, because I am printing explanatory lines a few times in the example. Namely:
>   
>   * `print("Position data array (first 14 frames):")` and
>   * `print("Shape data array (first 14 frames):")`.
> * one is due to commented out code for `# sphinx_gallery_thumbnail_number = 2`, but this is the syntax to change the default thumbnail shown in the example. We are using the same approach in [another example](https://github.com/neuroinformatics-unit/movement/blob/85904d9c2a4fd65f1e78d046f9f869ba83ee88af/examples/compute_polar_coordinates.py#L304) too.
> 
> I think for clarity it makes sense to keep the duplicate lines as is and not define a separate variable, but lmk if you disagree.
> 
> I cannot retrieve the reports from previous tutorials anymore (examples [here](https://github.com/neuroinformatics-unit/movement/pull/180#issuecomment-2129496480) and [here](https://github.com/neuroinformatics-unit/movement/pull/135#issuecomment-2007119996)) but it seems in those cases we skipped three issues too.

Regarding the sonarclous issues, I agree that you can ignore them in this case. We don't mind some repetition in the examples (it's often a good thing, pedagogically).

File: examples/load_and_reindex_bboxes.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+print(ds.time)
+
+# %%
+# Inspect associated video
+# --------------------------------
+# The video associated to the data contains all 168 frames.
+
+video_path = dataset_dict["video"]
+
+video = sio.load_video(video_path)
```
[the remark this answers is not in the document]
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py:68]: yeah I wanted to say in group meeting today but forgot 😅

File: examples/load_and_reindex_bboxes.py [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+print(f"Number of frames: {n_frames}")
+print(f"Frame size: {width}x{height}")
+print(f"Number of channels: {channels}")
+
+
+# %%
+# We can plot the data over the corresponding video frames to
+# visualise the bounding boxes around the tracked crab.
+#
+# Let's focus on the first 15 frames of the video, and plot the annotated
```
[the remark this answers is not in the document]
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py:80]: yeah, I did 15 to have at least three different original samples to check, but can do with 2!

File: examples/load_and_reindex_bboxes.py [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+            confidence = ds.confidence.sel(
+                time=frame, individuals=individual
+            ).data
+            writer.writerow(
+                [frame, individual, x, y, width, height, confidence]
+            )
+
+# %%
+# Remove the output file
+# ----------------------
```
[the remark this answers is not in the document]
Comment (lochhh) [reply examples/load_and_reindex_bboxes.py:333]: Clean-up 😁

File: examples/load_and_reindex_bboxes.py [hunk 14]
```diff
@@ -0,0 +1,336 @@
+"""Load and reindex bounding boxes tracks
+==========================================
+
+Load an example dataset of bounding boxes' tracks and reindex
```
[the remark this answers is not in the document]
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py:4]: I found two files in which we use `bounding boxes' tracks` (here and in `docs/source/getting_started/movement_dataset.md`) and everywhere else we use `bounding boxes tracks`.

Just to confirm, would you like me to change all instances to `bounding boxes tracks`? If so, I'll probably do that in a separate PR for clarity.

File: examples/load_and_reindex_bboxes.py (hunk 9 above)
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py:302]: what is the name of this extension? 👀

File: examples/load_and_reindex_bboxes.py (hunk 1 above)
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py:2]: I think this is way clearer, thanks!

File: examples/load_and_reindex_bboxes.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (niksirbi) [reply examples/load_and_reindex_bboxes.py:4]: > Just to confirm, would you like me to change all instances to bounding boxes tracks? If so, I'll probably do that in a separate PR for clarity.

Yes, that would be my preference, but happy to be overruled on if you disagree.

File: examples/load_and_reindex_bboxes.py (hunk 9 above)
Comment (niksirbi) [reply examples/load_and_reindex_bboxes.py:303]: Actually I'm not sure which one of the linting-related extensions flagged this, I have too many, and the warning itself didn't include a "source".

File: examples/load_and_reindex_bboxes.py [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+# ``shape`` arrays by taking the last valid value in time. In this way, a
+# box's position and shape stay constant if for a current frame the box
+# has no annotation defined.
+
+ds_ff = ds.reindex(
+    {"time": list(range(ds.time[-1].item()))},
+    method="ffill",  # propagate last valid index value forward
+)
+
+# check the first 14 frames of the data
```
[the remark this answers is not in the document]
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py:168]: It is somewhat related to the 15 above (I didn't realise I chose different values here actually), the aim is to show how the filling works for 3 original data points.

I guess one could argue that the minimum number of samples to show that it works as expected for 3 original data points, is to show 12 samples? But probably it's enough to show it for 2 original data points.

I can instead plot it though, that is probably nicer 🤔

File: examples/load_and_reindex_bboxes.py [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+        s=5,
+        color="white",
+    )
+
+    ax.set_title(f"Frame {ds.time[data_idx].item()}")
+    ax.set_xlabel("x (pixles)")
+    ax.set_ylabel("y (pixels)")
+    ax.set_xlabel("")
+
+fig.tight_layout()
```
[the remark this answers is not in the document]
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py]: yeah using `.sel` makes things easier here, thanks for the tip.

I am sometimes put off by `.sel` / `.isel` cause they are quite verbose 🫠

File: examples/load_and_reindex_bboxes.py (hunk 6 above)
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py]: agree, I found the wording for this example to be quite awkward myself too 😅

File: examples/load_and_reindex_bboxes.py (hunk 7 above)
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py]: I used `ds_filled` (`ds_` seemed to close to the loop variable `ds_i`)

File: examples/load_and_reindex_bboxes.py (hunk 10 above)
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py]: I think it's probably nice to keep it simple in this case and stick to frames

File: examples/load_and_reindex_bboxes.py (hunk 8 above)
Comment (sfmig) [reply examples/load_and_reindex_bboxes.py]: I made the markers a bit bigger for visibility

File: examples/load_and_upsample_bboxes.py [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+# %%
+# We can verify with a plot that the empty values have been filled in
+# using the last valid value in time.
+#
+# For this we define a convenience function to plot the x and y coordinates
+# of the ``position`` and  ``shape`` arrays, for the input dataset and for
+# a filled one.
+
+
+def plot_position_and_shape_xy_coords(ds_input_data, ds_filled, color_filled):
```
Comment (niksirbi) [inline examples/load_and_upsample_bboxes.py:184]: Nice! This plot illustrates the various interpolation methods in a much more clear way!

File: examples/load_and_upsample_bboxes.py [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+    """Compare the x and y coordinates of the position and shape arrays in time
+    for the input and filled datasets.
+    """
+    _, axs = plt.subplots(2, 2, figsize=(8, 6))
```
Comment (niksirbi) [inline examples/load_and_upsample_bboxes.py:188]: Let's make the figure a bit wider, on my screen there is barely any horizontal white space between the subplots.
```suggestion
    _, axs = plt.subplots(2, 2, figsize=(9, 6))
```

File: examples/load_and_reindex_bboxes.py (hunk 9 above)
Comment (niksirbi) [reply examples/load_and_reindex_bboxes.py]: I found the extension: https://github.com/nhoizey/vscode-gremlins

File: examples/load_and_upsample_bboxes.py (hunk 18 above)
Comment (niksirbi) [reply examples/load_and_upsample_bboxes.py:188]: or alternatively use `plt.tight_layout()`

Comment (niksirbi) [reply examples/load_and_upsample_bboxes.py:188]: Here's a screenshot so you know what I mean: 

Screenshot 2024-12-05 at 18 24 19

**Review by niksirbi — APPROVED** [review]
Just 1 comment @sfmig, and after that it's good to merge!

I really like the new plots you've added now.
